<head>
<title>PROSITE: PDOC00640 (documentation file)</title>
</head>
<body bgcolor=#FFFFFF text=#000000>
<hr SIZE=2 shade><center><table width="70%" border=2 cellspacing=0 cellpadding=2 align=center><tr>
<td bgcolor="#FFFFFF" align=center> <b><a TARGET="_top" href="/">ExPASy Home page</a></b></td>
<td bgcolor="#FF9999" align=center> <b><a TARGET="_top" href="/sitemap.html">Site Map</a></b></td>
<td bgcolor="#C0EAC0" align=center> <b><a TARGET="_top" href="/ExpasyHunt/">Search ExPASy</a></b></td>
<td bgcolor="#CCCCFF" align=center> <b><a TARGET="_top" href="/contact.html">Contact us</a></b></td>
<td bgcolor="#33CCFF" align=center> <b><a TARGET="_top" href="/prosite/">PROSITE</a></b></td>
</tr></table></center><center><table border=1><tr><td>Mirror sites:</td><td><A HREF="http://expasy.proteome.org.au/cgi-bin/get-prodoc-entry?PDOC00640">Australia</A></td> <td><A HREF="http://expasy.cbr.nrc.ca/cgi-bin/get-prodoc-entry?PDOC00640">Canada</A></td> <td><A HREF="http://expasy.pku.edu.cn/cgi-bin/get-prodoc-entry?PDOC00640">China</A></td> <td><A HREF="http://expasy.nhri.org.tw/cgi-bin/get-prodoc-entry?PDOC00640">Taiwan</A></td> </tr></table></center><hr SIZE=2 shade>

<h1>PROSITE: PDOC00640 (documentation)</h1>
<p><Address><A HREF="/cgi-bin/nicedoc.pl?PDOC00640">View entry in NiceSite format</A></Address>
<pre>
{PDOC00640}
<A HREF="/cgi-bin/get-prosite-entry?PS00812">{PS00812; GLYCOSYL_HYDROL_F8}
</A>{BEGIN}
******************************************
* Glycosyl hydrolases family 8 signature *
******************************************

The microbial degradation  of cellulose and  xylans requires  several types of
enzymes such as endoglucanases (EC <A HREF= "/cgi-bin/get-enzyme-entry?3.2.1.4">3.2.1.4</A>),  cellobiohydrolases (EC <A HREF= "/cgi-bin/get-enzyme-entry?3.2.1.91">3.2.1.91</A>)
(exoglucanases), or xylanases (EC <A HREF= "/cgi-bin/get-enzyme-entry?3.2.1.8">3.2.1.8</A>) [1,2].  Fungi and bacteria produces
a spectrum of cellulolytic  enzymes (cellulases)  and  xylanases which, on the
basis of sequence similarities,  can be classified into families. One of these
families is known as the cellulase family D [3] or as  the glycosyl hydrolases
family 8  [4,<A HREF="http://www.expasy.ch/cgi-bin/lists?glycosid.txt">E1</A>].  The  enzymes  which  are  currently known to belong to this
family are listed below.

 - Acetobacter xylinum endonuclease cmcAX.
 - Bacillus strain KSM-330 acidic endonuclease K (Endo-K).
 - Cellulomonas josui endoglucanase 2 (celB).
 - Cellulomonas uda endoglucanase.
 - Clostridium cellulolyticum endoglucanases C (celcCC).
 - Clostridium thermocellum endoglucanases A (celA).
 - Erwinia chrysanthemi minor endoglucanase y (celY).
 - Bacillus circulans beta-glucanase (EC <A HREF= "/cgi-bin/get-enzyme-entry?3.2.1.73">3.2.1.73</A>).
 - Escherichia coli hypothetical protein yhjM.

The most conserved region in  these enzymes is  a stretch of about 20 residues
that contains  two conserved aspartate. The first asparatate is thought [5] to
act as the nucleophile in the catalytic mechanism. We have used this region as
a signature pattern.

-Consensus pattern: A-[ST]-D-[AG]-D-x(2)-[IM]-A-x-[SA]-[LIVM]-[LIVMG]-x-A-
                    x(3)-[FW]
                    [The first D is an active site residue]
-Sequences known to belong to this class detected by the pattern: ALL.
-Other sequence(s) detected in SWISS-PROT: NONE.

-Expert(s) to contact by email:
                   Henrissat B.; <A HREF= "/cgi-bin/mailform/exp?___________Henrissat_B.&Glycosyl_hydrolases_family_8_signature&bernie@afmb.cnrs-mrs.fr
">bernie@afmb.cnrs-mrs.fr
</A>
-Last update: November 1997 / Text revised.

[ 1] Beguin P.
     Annu. Rev. Microbiol. 44:219-248(1990).
[ 2] Gilkes N.R., Henrissat B., Kilburn D.G., Miller R.C. Jr., Warren R.A.J.
     Microbiol. Rev. 55:303-315(1991).
[ 3] Henrissat B., Claeyssens M., Tomme P., Lemesle L., Mornon J.-P.
     Gene 81:83-95(1989).
[ 4] Henrissat B.
     Biochem. J. 280:309-316(1991).
[ 5] Alzari P.M., Souchon H., Dominguez R.
     Structure 4:265-275(1996).
[E1] <A HREF="/cgi-bin/lists?glycosid.txt">http://www.expasy.ch/cgi-bin/lists?glycosid.txt</A>

+----------------------------------------------------------------------------+
| This PROSITE entry  is copyright  by the Swiss Institute of Bioinformatics |
| (SIB).  There are no restrictions on its use by non-profit institutions as |
| long as its  content  is  in  no  way  modified  and this statement is not |
| removed. Usage by and for commercial entities requires a license agreement |
| (See <A HREF="http://www.isb-sib.ch/announce/" >http://www.isb-sib.ch/announce/</A> or email to <A HREF="mailto:license@isb-sib.ch">license@isb-sib.ch</A>).      |
+----------------------------------------------------------------------------+
{END}
</PRE><hr SIZE=2 shade><center><table width="70%" border=2 cellspacing=0 cellpadding=2 align=center><tr>
<td bgcolor="#FFFFFF" align=center> <b><a TARGET="_top" href="/">ExPASy Home page</a></b></td>
<td bgcolor="#FF9999" align=center> <b><a TARGET="_top" href="/sitemap.html">Site Map</a></b></td>
<td bgcolor="#C0EAC0" align=center> <b><a TARGET="_top" href="/ExpasyHunt/">Search ExPASy</a></b></td>
<td bgcolor="#CCCCFF" align=center> <b><a TARGET="_top" href="/contact.html">Contact us</a></b></td>
<td bgcolor="#33CCFF" align=center> <b><a TARGET="_top" href="/prosite/">PROSITE</a></b></td>
</tr></table></center><center><table border=1><tr><td>Mirror sites:</td><td><A HREF="http://expasy.proteome.org.au/cgi-bin/get-prodoc-entry?PDOC00640">Australia</A></td> <td><A HREF="http://expasy.cbr.nrc.ca/cgi-bin/get-prodoc-entry?PDOC00640">Canada</A></td> <td><A HREF="http://expasy.pku.edu.cn/cgi-bin/get-prodoc-entry?PDOC00640">China</A></td> <td><A HREF="http://expasy.nhri.org.tw/cgi-bin/get-prodoc-entry?PDOC00640">Taiwan</A></td> </tr></table></center><hr SIZE=2 shade>

