<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
            "http://www.w3.org/TR/REC-html40/loose.dtd">
<HTML>
<HEAD><TITLE></TITLE>

<META http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<META name="GENERATOR" content="hevea 1.07">
</HEAD>
<BODY >
<!--HEVEA command line is: hevea Installation.tex -->
<!--HTMLHEAD-->
<!--ENDHTML-->
<!--PREFIX <ARG ></ARG>-->
<!--CUT DEF section 1 -->
<BR>
<BR>



<H1 ALIGN=center>Biopython Installation</H1>

<H3 ALIGN=center>Brad Chapman (chapmanb@uga.edu)</H3>
<!--TOC section Table of Contents-->

<H2>Table of Contents</H2><!--SEC END -->

<UL><LI>
<A HREF="#htoc1">1&nbsp;&nbsp;Purpose and Assumptions</A>
<LI><A HREF="#htoc2">2&nbsp;&nbsp;Installing Python</A>
<UL><LI>
<A HREF="#htoc3">2.1&nbsp;&nbsp;Installation on UNIX systems and Mac OS X</A>
<UL><LI>
<A HREF="#htoc4">2.1.1&nbsp;&nbsp;RPM and other Package Manager Installation</A>
</UL>
<LI><A HREF="#htoc5">2.2&nbsp;&nbsp;Installation on Windows</A>
<LI><A HREF="#htoc6">2.3&nbsp;&nbsp;Installation on older Macintoshes</A>
</UL>
<LI><A HREF="#htoc7">3&nbsp;&nbsp;Installing Biopython dependencies</A>
<UL><LI>
<A HREF="#htoc8">3.1&nbsp;&nbsp;mxTextTools</A>
<UL><LI>
<A HREF="#htoc9">3.1.1&nbsp;&nbsp;UNIX and Mac OS X systems</A>
<LI><A HREF="#htoc10">3.1.2&nbsp;&nbsp;Windows systems</A>
<LI><A HREF="#htoc11">3.1.3&nbsp;&nbsp;Making sure it installed correctly</A>
</UL>
<LI><A HREF="#htoc12">3.2&nbsp;&nbsp;Numerical Python</A>
<UL><LI>
<A HREF="#htoc13">3.2.1&nbsp;&nbsp;UNIX and Mac OS X systems</A>
<LI><A HREF="#htoc14">3.2.2&nbsp;&nbsp;Windows systems</A>
<LI><A HREF="#htoc15">3.2.3&nbsp;&nbsp;Making sure it installed correctly</A>
</UL>
<LI><A HREF="#htoc16">3.3&nbsp;&nbsp;ReportLab (optional)</A>
<UL><LI>
<A HREF="#htoc17">3.3.1&nbsp;&nbsp;UNIX and Mac OS X systems</A>
<LI><A HREF="#htoc18">3.3.2&nbsp;&nbsp;Windows systems</A>
<LI><A HREF="#htoc19">3.3.3&nbsp;&nbsp;Making sure it installed correctly</A>
</UL>
<LI><A HREF="#htoc20">3.4&nbsp;&nbsp;Database Access (MySQLdb, ...) (optional)</A>
</UL>
<LI><A HREF="#htoc21">4&nbsp;&nbsp;Installing Biopython</A>
<UL><LI>
<A HREF="#htoc22">4.1&nbsp;&nbsp;Obtaining Biopython
</A>
<LI><A HREF="#htoc23">4.2&nbsp;&nbsp;Installing from source on UNIX and Mac OS X</A>
<UL><LI>
<A HREF="#htoc24">4.2.1&nbsp;&nbsp;Installation on FreeBSD</A>
</UL>
<LI><A HREF="#htoc25">4.3&nbsp;&nbsp;Installing on UNIX systems using RPMs</A>
<LI><A HREF="#htoc26">4.4&nbsp;&nbsp;Installing with a Windows Installer</A>
<LI><A HREF="#htoc27">4.5&nbsp;&nbsp;Installing from source on Windows</A>
<LI><A HREF="#htoc28">4.6&nbsp;&nbsp;Installing on Macintosh</A>
<UL><LI>
<A HREF="#htoc29">4.6.1&nbsp;&nbsp;Pre-Built Install</A>
<LI><A HREF="#htoc30">4.6.2&nbsp;&nbsp;Non Pre-Built Installation</A>
</UL>
</UL>
<LI><A HREF="#htoc31">5&nbsp;&nbsp;Making sure everything worked</A>
<LI><A HREF="#htoc32">6&nbsp;&nbsp;Notes for installing with non-administrator permissions</A>
</UL>

<!--TOC section Purpose and Assumptions-->

<H2><A NAME="htoc1">1</A>&nbsp;&nbsp;Purpose and Assumptions</H2><!--SEC END -->

This document describes installing Biopython on your computer. To make
things as simple as possible, it basically assumes you have nothing
related to Python or Biopython on your computer and want to end up with
a working installation of Biopython when you are finished following
through this documentation. <BR>
<BR>
Biopython should work on just any operating system where Python works,
so these instructions contain directions for installation on UNIX/Linux
, Windows and Macintosh machines. The directions assume 
that you have permission to install programs on the machine
(root access on UNIX and Administrator privileges on Windows or Mac
machines). While it is certainly possible to install things without
these privileges, this is a serious pain and all the tedious workarounds
aren't something that I'll go into very much in this documentation.<BR>
<BR>
With all this said, hopefully these directions will make it
straightforward to get Biopython installed on your machine so you can
begin using it as quick as possible.<BR>
<BR>
<!--TOC section Installing Python-->

<H2><A NAME="htoc2">2</A>&nbsp;&nbsp;Installing Python</H2><!--SEC END -->

Python is a interpreting, interactive object-oriented programming
language and the home for all things python is
<A HREF="http://www.python.org"><TT>http://www.python.org</TT></A>. Presumedly you have some idea of
python and what it can do if you are interested in Biopython, but if not
the python website contains tons of documentation and reasons to learn
to program in python.<BR>
<BR>
Biopython is designed to work with Python 2.3 or later. With python, the
general rule of thumb is to keep yourself using the latest version, as
the development process is very clean and new releases are quite stable.
Upgrading bug-fix releases (for example. 2.3.1 to 2.3.2) 
is incredibly easy and won't require any re-installation of libraries.
Upgrading between versions (2.3 to 2.4) is more time consuming since you
need to re-install all libraries you have added to python.<BR>
<BR>
Let's get started with installation on various platforms.<BR>
<BR>
<!--TOC subsection Installation on UNIX systems and Mac OS X-->

<H3><A NAME="htoc3">2.1</A>&nbsp;&nbsp;Installation on UNIX systems and Mac OS X</H3><!--SEC END -->

First, you should go the main python web site and head over to the information
page for the latest python release. At the time of this writing the
latest stable python release is 2.4, which is available from
<A HREF="http://www.python.org/2.4/"><TT>http://www.python.org/2.4/</TT></A>. This page contains links
to all released files for the given release. For UNIX, we'll want to use
the tarred and gzipped file, which is called <CODE>Python-2.4.tgz</CODE> at
the time of this writing.<BR>
<BR>
Download this file and then unpack it with the following commands:
<PRE>
$ gunzip Python-2.4.tgz 
$ tar -xvpf Python-2.4.tar 
</PRE>
Then enter into the created directory:
<PRE>
$ cd Python-2.4
</PRE>
Now, start the build process by configuring everything to your system:
<PRE>
$ ./configure
</PRE>
Build all of the files with:
<PRE>
$ make
</PRE>
Finally, you'll need to have root permissions on the system and then
install everything:
<PRE>
# make install
</PRE>
If there were no errors and everything worked correctly, you should now
be able to type <CODE>python</CODE> at a command prompt and enter into the
python interpreter:
<PRE>
$ python
Python 2.4 (#1, Dec  5 2004, 20:47:03)
[GCC 3.3.3] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt;
</PRE>
<!--TOC subsubsection RPM and other Package Manager Installation-->

<H4><A NAME="htoc4">2.1.1</A>&nbsp;&nbsp;RPM and other Package Manager Installation</H4><!--SEC END -->

There are a multitude of package manager systems out there for which
python is available. One popular one is the RPM (RedHat Package Manager)
system. Each of these package managing systems has its own quirks and
tricks and I certainly can't pretend to understand them all so I won't
try to describe them all here.<BR>
<BR>
However, there is one general point which it is important to remember
when installing from any of these systems: you need to download and
install the development packages for python. A number of distributions
contain a "basic" python which contains libraries and enough stuff to
run simple python programs. However, they do not contain the python
libraries necessary to build third-party python applications (like
Biopython and it's dependencies). You'll need to install these libraries
and header files, which are often found in a separate package called
<CODE>python-devel</CODE> or something similar. <BR>
<BR>
<!--TOC subsection Installation on Windows-->

<H3><A NAME="htoc5">2.2</A>&nbsp;&nbsp;Installation on Windows</H3><!--SEC END -->
 
Installation on Windows is most easily done using handy windows
installers. As described above in the UNIX section, you should go to the
webpage for the current stable version of Python to download this
installer. At the current time, you'd go to
<A HREF="http://www.python.org/2.4/"><TT>http://www.python.org/2.4/</TT></A> and download
<CODE>Python-2.4.msi</CODE>. <BR>
<BR>
The installer is an executable program, so you only need to double click
it to run it. Then just follow the friendly instructions. On all newer Windows
machines you'll need to have Administrator privileges to do this
installation.<BR>
<BR>
<!--TOC subsection Installation on older Macintoshes-->

<H3><A NAME="htoc6">2.3</A>&nbsp;&nbsp;Installation on older Macintoshes</H3><!--SEC END -->

Mac OS X can readily compile the python distribution in the way
described above for UNIX systems, but earlier versions of the Macintosh
operating system aren't nearly as easy to work with. For these versions,
the best place to go is the Macintosh page on the python site:
<A HREF="http://www.python.org/download/download_mac.html"><TT>http://www.python.org/download/download_mac.html</TT></A>. This
site links to the MacPython pages, which contain installers for
Macintosh.<BR>
<BR>
The Macintosh installer is as simple to use at the Windows installers.
You download the appropriate file (<CODE>MacPython222active.bin</CODE> at the
current time), and then unpack it with StuffIt Expander. You then just
need to double click on the resulting graphical installer and follow the
easy instructions.<BR>
<BR>
<!--TOC section Installing Biopython dependencies-->

<H2><A NAME="htoc7">3</A>&nbsp;&nbsp;Installing Biopython dependencies</H2><!--SEC END -->

Once python is installed, the next step is getting the dependencies
for Biopython installed. Since not all functionality is included in the
main python installation, Biopython needs some support libraries to save
us a lot of work re-writing code that already exists. We try to keep
as few dependencies as possible to make installation as easy as
possible.<BR>
<BR>
<!--TOC subsection mxTextTools-->

<H3><A NAME="htoc8">3.1</A>&nbsp;&nbsp;mxTextTools</H3><!--SEC END -->

This is the most important Biopython dependency as it is used
extensively in the internals of a number of parsers. You absolutely want
to install this if you want to get any sort of serious use out of
Biopython.<BR>
<BR>
mxTextTools is available along with the entire mx-base system (which
contains a number of other useful utilities as well) and is available
for download at:
<A HREF="http://www.lemburg.com/files/python/mxTextTools.html"><TT>http://www.lemburg.com/files/python/mxTextTools.html</TT></A>.<BR>
<BR>
<!--TOC subsubsection UNIX and Mac OS X systems-->

<H4><A NAME="htoc9">3.1.1</A>&nbsp;&nbsp;UNIX and Mac OS X systems</H4><!--SEC END -->

For UNIX and UNIX-like systems you should download the <CODE>tar.gz</CODE>
file from the page listed above. At the current time, this is
<CODE>egenix-mx-base-2.0.6.tar.gz</CODE>.<BR>
<BR>
Once you download this, unpack it and change into the created directory:
<PRE>
$ gunzip egenix-mx-base-2.0.6.tar.gz 
$ tar -xvpf egenix-mx-base-2.0.6.tar
$ cd egenix-mx-base-2.0.6
</PRE>
To build it, use the standard python build procedure:
<PRE>
$ python setup.py build
</PRE>
Then become root, and install it, again using the standard python
mechanism:
<PRE>
$ python setup.py install
</PRE>
<!--TOC subsubsection Windows systems-->

<H4><A NAME="htoc10">3.1.2</A>&nbsp;&nbsp;Windows systems</H4><!--SEC END -->

For Windows operating systems, you should download the Windows installer
for the version of python you are running. At the current time, this
would be: <CODE>egenix-mx-base-2.0.6.win32-py2.4.exe</CODE>.<BR>
<BR>
This is a standard graphical installer, so after download double click
it and follow the instructions and it should install with no problem.
You'll have to have Administrator privileges to do this install, as with
python itself.<BR>
<BR>
<!--TOC subsubsection Making sure it installed correctly-->

<H4><A NAME="htoc11">3.1.3</A>&nbsp;&nbsp;Making sure it installed correctly</H4><!--SEC END -->

If you've installed mxTextTools correctly, you should be able to fire up
your python interpreter and import it with no errors:
<PRE>
$ python
Python 2.4 (#1, Dec  5 2004, 20:47:03)
[GCC 3.3.3] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt; from mx import TextTools
&gt;&gt;&gt;
</PRE>
<!--TOC subsection Numerical Python-->

<H3><A NAME="htoc12">3.2</A>&nbsp;&nbsp;Numerical Python</H3><!--SEC END -->

The Numerical Python distribution (also known an Numeric or Numpy) is a
fast implementation of arrays and associated array functionality. This
is important for a number of Biopython modules that deal with
number processing. The main web site for Numeric is:
<A HREF="http://sourceforge.net/projects/numpy"><TT>http://sourceforge.net/projects/numpy</TT></A> and downloads are
available from:
<A HREF="http://sourceforge.net/project/showfiles.php?group_id=1369&package_id=1351"><TT>http://sourceforge.net/project/showfiles.php?group_id=1369&amp;package_id=1351</TT></A>.<BR>
<BR>
<!--TOC subsubsection UNIX and Mac OS X systems-->

<H4><A NAME="htoc13">3.2.1</A>&nbsp;&nbsp;UNIX and Mac OS X systems</H4><!--SEC END -->

As with mxTextTools, you should download the <CODE>tar.gz</CODE> file. Version
23.8 of Numerical Python (<CODE>Numeric-23.8.tar.gz</CODE>) compiles out of the box.
The build process is exactly the same as with mxTextTools:
<PRE>
$ gunzip Numeric-23.8.tar.gz 
$ tar -xvpf Numeric-23.8.tar
$ cd Numeric-23.8
$ python setup.py build
</PRE>
Once it is built, you should become root, and then install it:
<PRE>
$ python setup.py install
</PRE>
One important note if you use an RPM-based system and not installing
from source as described above: you need to also
install the Numeric header files which are not included with some
Numeric packages. As with the main python distribution, this means
you'll need to look for something like <CODE>python-numeric-devel</CODE> 
and make sure to install this as well as the basic Numeric package.<BR>
<BR>
<!--TOC subsubsection Windows systems-->

<H4><A NAME="htoc14">3.2.2</A>&nbsp;&nbsp;Windows systems</H4><!--SEC END -->

Once again, Windows installers are available for Numeric so you should
follow the now-standard procedure of downloading the installer
(<CODE>Numeric-23.8.win32-py2.4.exe</CODE> at the current time), double
clicking it and then following the installation instructions. As before,
you will need to have administrator permissions to do this.<BR>
<BR>
<!--TOC subsubsection Making sure it installed correctly-->

<H4><A NAME="htoc15">3.2.3</A>&nbsp;&nbsp;Making sure it installed correctly</H4><!--SEC END -->

To make sure everything went okay during the install, fire up the python
interpreter and ensure you can import Numeric without any errors:
<PRE>
$ python
Python 2.4 (#1, Dec  5 2004, 20:47:03)
[GCC 3.3.3] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt; from Numeric import *
&gt;&gt;&gt;
</PRE>
<!--TOC subsection ReportLab (optional)-->

<H3><A NAME="htoc16">3.3</A>&nbsp;&nbsp;ReportLab (optional)</H3><!--SEC END -->

The ReportLab package is a library for generating PDF documents. It is
used in the Biopython Graphics modules, which contains basic
functionality for drawing biological objects like chromosomes. If you
are not planning on using this, installing ReportLab is not necessary.
ReportLab in itself is very useful for a number of tasks besides just
Biopython, so you may want to check out
<A HREF="http://www.reportlab.org"><TT>http://www.reportlab.org</TT></A> before making your decision.<BR>
<BR>
The main download page for ReportLab is
<A HREF="http://www.reportlab.org/downloads.html"><TT>http://www.reportlab.org/downloads.html</TT></A>. The ReportLab
company has some commercial products as well, but just scroll down their
page to the Open Source software section for the base ReportLab
downloads.<BR>
<BR>
<!--TOC subsubsection UNIX and Mac OS X systems-->

<H4><A NAME="htoc17">3.3.1</A>&nbsp;&nbsp;UNIX and Mac OS X systems</H4><!--SEC END -->

For UNIX installs, you should download the tarred and gzipped version of
the ReportLab distribution. At the time of this writing, this is called
<CODE>ReportLab_1_20.tgz</CODE>. First, unpack the distribution and change
into the created directory:
<PRE>
$ gunzip ReportLab_1_20.tgz 
$ tar -xvpf ReportLab_1_20.tar
$ cd reportlab_1_20/
</PRE>
Once again, ReportLab uses the standard python installation system which
you are probably feeling really comfortable with by now. So, first build
the package:
<PRE>
$ python setup.py build
</PRE>
Now become root, and install it:
<PRE>
$ python setup.py install
</PRE>
<!--TOC subsubsection Windows systems-->

<H4><A NAME="htoc18">3.3.2</A>&nbsp;&nbsp;Windows systems</H4><!--SEC END -->

ReportLab does not have a graphical windows installer like the other
Biopython requirements. Luckily, it doesn't require any compilation
steps to work properly, so the installation is still quite easy. <BR>
<BR>
First, download the zipped distribution from the download site listed
above. At the current time this is called <CODE>ReportLab_1_20.zip</CODE>. You
can also download the tarred/gzipped file (with a <CODE>.tgz</CODE>
extension), but Windows handles zipped files better.<BR>
<BR>
Secondly, unzip the downloaded file. WinZip is a common freely available
program for doing this
(<A HREF="http://www.winzip.com/ddchomea.htm"><TT>http://www.winzip.com/ddchomea.htm</TT></A>). The unzip process
should create a <CODE>reportlab</CODE> directory.<BR>
<BR>
Finally, drag the created <CODE>reportlab</CODE> directory to the standard
directory for Python extensions. On current versions of python with a
standard installation this would be something like 
<CODE>C:/Python24/Lib/site-packages</CODE>. All you have to do is drag it over
and you should be all set. Nice and easy.<BR>
<BR>
<!--TOC subsubsection Making sure it installed correctly-->

<H4><A NAME="htoc19">3.3.3</A>&nbsp;&nbsp;Making sure it installed correctly</H4><!--SEC END -->

If reportlab is installed correctly, you should be able to do the
following:
<PRE>
$ python
Python 2.4 (#1, Dec  5 2004, 20:47:03)
[GCC 3.3.3] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt; from reportlab.graphics import renderPDF
&gt;&gt;&gt;
</PRE>
Depending on your version of python and what you have installed, you may
get the following warning message: 
<CODE>Warn: Python Imaging Library not available</CODE>. This isn't anything
to worry about since the Biopython parts that use ReportLab will work
just fine without it.<BR>
<BR>
<!--TOC subsection Database Access (MySQLdb, ...) (optional)-->

<H3><A NAME="htoc20">3.4</A>&nbsp;&nbsp;Database Access (MySQLdb, ...) (optional)</H3><!--SEC END -->

The MySQLdb package is a library for accessing MySQL databases. It is
used in the Biopython GFF module, which allows access to databases
created from GFF feature tables. Biopython also includes an accessory
module, DocSQL, which provides a convenient interface to MySQLdb. 
If you are not planning on using Bio.GFF or Bio.DocSQL, installing
MySQLdb is not necessary.<BR>
<BR>
Additionally, both MySQLdb and psycopg (a PostgreSQL database adaptor)
can be used for accessing BioSQL databases through Biopython. Again if
you are not going to use BioSQL, there shouldn't be any need to install
these modules.<BR>
<BR>
Installation instructions for MySQLdb and psycopg are included in the
BioSQL documentation, which is available from
<A HREF="http://www.biopython.org/docs/biosql/python_biosql_basic.html"><TT>http://www.biopython.org/docs/biosql/python_biosql_basic.html</TT></A> and
<A HREF="http://www.biopython.org/docs/biosql/python_biosql_basic.pdf"><TT>http://www.biopython.org/docs/biosql/python_biosql_basic.pdf</TT></A>.<BR>
<BR>
<!--TOC section Installing Biopython-->

<H2><A NAME="htoc21">4</A>&nbsp;&nbsp;Installing Biopython</H2><!--SEC END -->

<!--TOC subsection Obtaining Biopython
-->

<H3><A NAME="htoc22">4.1</A>&nbsp;&nbsp;Obtaining Biopython
</H3><!--SEC END -->

Biopython's internet home is at, naturally enough, 
<A HREF="http://www.biopython.org"><TT>http://www.biopython.org</TT></A>. This is the home of all things 
Biopython, so it is the best place to start looking around. 
You have two choices for obtaining Biopython:
<OL type=1><LI>Release code -- We made available releases on the download page 
(<A HREF="http://www.biopython.org/download/"><TT>http://www.biopython.org/download/</TT></A>). 
The releases are also available both as source and as installers 
(windows installers right now), so you have some choices to pick from 
on releases if you prefer not to deal with source code directly.<BR>
<BR>
<LI>CVS -- The current working copy of the Biopython sources is always
available via CVS (Concurrent Versions Systems --
<A HREF="http://www.cvshome.org/"><TT>http://www.cvshome.org/</TT></A>). Concise instructions for
accessing this copy are available at
<A HREF="http://cvs.biopython.org"><TT>http://cvs.biopython.org</TT></A>. CVS is normally quite stable
but there is always the caveat that the code there is under
development.</OL>
Based on which way you choose, you'll need to follow one of the following installation options. Read on for the platform you are working on.<BR>
<BR>
<!--TOC subsection Installing from source on UNIX and Mac OS X-->

<H3><A NAME="htoc23">4.2</A>&nbsp;&nbsp;Installing from source on UNIX and Mac OS X</H3><!--SEC END -->

<A NAME="sec:unix_install"></A>
Biopython uses Distutils, the standard python installation package, for
its installation. If you read the install instructions above you are
already quite familiar with its workings. Distutils comes standard with 
Python 1.6 and beyond.<BR>
<BR>
Now that we've got what we need, let's get into the installation:
<OL type=1><LI>First you need to unpack the distribution. If you got the CVS version, you are all set to go and can skip on ahead. Otherwise, you'll need to unpack it. On UN*X machines, a tar.gz package is provided, which you can unpack with <CODE>tar -xzvpf biopython-X.X.tar.gz</CODE>. A zip file is also provided for other platforms.<BR>
<BR>
<LI>Now that everything is unpacked, move into the <CODE> biopython*</CODE> directory (this will just be <CODE>biopython</CODE> for CVS users, and will be <CODE>biopython-X.X</CODE> for those using a packaged download). <BR>
<BR>
<LI>Now you are ready for your one step install -- <CODE>python setup.py install</CODE>. This performs the default install, and will put Biopython into the <CODE>site-packages</CODE> directory of your python library tree (on my machine this is <CODE>/usr/local/lib/python2.4/site-packages</CODE>). You will have to have permissions to write to this directory, so you'll need to have root access on the machine.
<OL type=a><LI>This install requires that you have the python source available. You can check this by looking for <CODE>Python.h</CODE> and <CODE>config.h</CODE> in some place like <CODE>/usr/local/include/python2.4</CODE>. If you installed python with RPMs or 
some other packaging system, this means you'll also have to install the
header files. This requires installing the python development libraries
as well (normally called something like <CODE>python-devel-2.4.rpm</CODE>).<BR>
<BR>
<LI>The distutils setup process allows you to do some customization of your install so you don't have to stick everything in the default location (in case you don't have write permissions there, or just want to test Biopython out). You have quite a few choices, which are covered in detail in the distutils installation manual (<A HREF="http://www.python.org/sigs/distutils-sig/doc/inst/inst.html"><TT>http://www.python.org/sigs/distutils-sig/doc/inst/inst.html</TT></A>), specifically in the Alternative installation section. For example, to install Biopython into your home directory, you need to type <CODE>python setup.py install --home=$HOME</CODE>. This will install the package into someplace like <CODE>$HOME/lib/python2.4/site-packages</CODE>. You'll need to subsequently modify the <CODE>PYTHONPATH</CODE> environmental variable to include this directory so python will be able to find the installation.</OL><BR>
<BR>
<LI>That's it! Biopython is installed. Wasn't that easy? Now let's check and make sure it worked properly. Skip on ahead to section&nbsp;<A HREF="#sec:is_working">5</A>.</OL>
<!--TOC subsubsection Installation on FreeBSD-->

<H4><A NAME="htoc24">4.2.1</A>&nbsp;&nbsp;Installation on FreeBSD</H4><!--SEC END -->

Johann Visagie has been kind enough to create (and keep updated) a FreeBSD port of Biopython. Thanks to the wonders of the ports system, this means that all you need to do to install Biopython on FreeBSD is do the following as root:
<PRE>
# cd /usr/ports/biology/py-biopython
# make install
</PRE>
And voila! It's installed. <BR>
<BR>
If you want more information on FreeBSD and things, Johann has written a nice primer for his FreeBSD EMBOSS port. This has lots of generally useful information, such as how to keep your ports tree up to date. If you are new to FreeBSD, you should definitely check it out at <A HREF="ftp://ftp.no.embnet.org/pub/EMBOSS-extras/EMBOSS-FreeBSD-HOWTO.txt"><TT>ftp://ftp.no.embnet.org/pub/EMBOSS-extras/EMBOSS-FreeBSD-HOWTO.txt</TT></A>.<BR>
<BR>
<!--TOC subsection Installing on UNIX systems using RPMs-->

<H3><A NAME="htoc25">4.3</A>&nbsp;&nbsp;Installing on UNIX systems using RPMs</H3><!--SEC END -->

Warning. Right now we're not making RPMs for biopython (because I
stopped using an RPM system, basically). If anyone wants to pick this
up, or feels especially strongly that they'd like RPMs, please let us
know.<BR>
<BR>
To simplify things for people running RPM-based systems, biopython can
also be installed via the RPM system. Additionally, this saves the 
necessity of having a C compiler to install biopython. <BR>
<BR>
Installing Biopython from a RPM package should be much the same process as used for other RPMs. If you need general information about how RPMs work, the best place to go is <A HREF="http://www.rpm.org"><TT>http://www.rpm.org</TT></A>.<BR>
<BR>
To install it, you should just need to do:
<PRE>
rpm -i your_biopython.rpm
</PRE>
To see what you installed try doing <CODE>rpm -qpl your_biopython.rpm</CODE> which will list all of the installed files.<BR>
<BR>
RPMs do not install the documentation, tests, or example code, so you might want to also grab a source distribution, so you can use these resources (and also look at the source code if you want to).<BR>
<BR>
<!--TOC subsection Installing with a Windows Installer-->

<H3><A NAME="htoc26">4.4</A>&nbsp;&nbsp;Installing with a Windows Installer</H3><!--SEC END -->

Installing things on Windows with the installer should be really easy (hey, that's why they've got graphical installers, right?). You should just need to download the <CODE>Biopython-version.exe</CODE> installer from biopython web site. Then you just need to double click and voila, a nice little installer will come up and you can stick the libraries where you need to. No need for a C compiler or anything fancy. You will need to have Administrator privileges on the machine to do the installation.<BR>
<BR>
This does not install the documentation, tests, example code or source code, so it is probably also a good idea to download the zip file containing this so you can test your installation and learn how to use it.<BR>
<BR>
<!--TOC subsection Installing from source on Windows-->

<H3><A NAME="htoc27">4.5</A>&nbsp;&nbsp;Installing from source on Windows</H3><!--SEC END -->

<A NAME="sec:windows_install"></A>
This section deals with installing the source (i.&nbsp;e.&nbsp;from CVS or from a source zip file) on a Windows machine. Much of the information from the UNIX install applies here, so it would be good to read section&nbsp;<A HREF="#sec:unix_install">4.2</A> before starting. Also, a little warning -- I (Brad) am writing these instructions based on very limited experience with Windows; I am basically a UNIX geek. So if you know more about Windows and want to add/correct things in this section, please feel let us know!<BR>
<BR>
I have successfully managed to use distutils to compile Biopython with Borland's free C++ compiler (available from <A HREF="http://www.inprise.com/bcppbuilder/freecompiler/"><TT>http://www.inprise.com/bcppbuilder/freecompiler/</TT></A>). It should also be possible with other Distutils supported compilers (please provide info if you've done this!).
<OL type=1><LI>
 Borland C++ compiler
<UL><LI>
 First you need to install and setup Borland C++. There are instructions on the Borland page and on the web, so you should follow these.<BR>
<BR>
<LI>Now you have to get python, which is compiled with a Microsoft compiler, able to live happily with Borland compiled extensions. Gordon Williams has an excellent page describing doing this (<A HREF="http://www.cyberus.ca/~g_will/pyExtenDL.shtml"><TT>http://www.cyberus.ca/~g_will/pyExtenDL.shtml</TT></A>), which is where I learned everything I know. Basically, what I did was run the Borland supplied tool <CODE>COFF2OMF</CODE> on the python20.lib file:
<PRE>
&gt; COFF2OMF python20.lib python20-borland.lib
</PRE>
Then I just renamed <CODE>python20.lib</CODE> to <CODE>python20-orig.lib</CODE> and renamed <CODE>python20-borland.lib</CODE> to <CODE>python20.lib</CODE>, so that distutils will link against this ``Borland friendly'' python library.<BR>
<BR>
<LI>Now that that is ready, we are at the easy part -- using distutils to build it. All I did was:
<PRE>
&gt; python setup.py build --compiler=bcpp
&gt; python setup.py install
</PRE>
and voila!, it's installed. </UL></OL>
Now that you've got everything installed, skip on ahead to section&nbsp;<A HREF="#sec:is_working">5</A> to make sure everything worked.<BR>
<BR>
<!--TOC subsection Installing on Macintosh-->

<H3><A NAME="htoc28">4.6</A>&nbsp;&nbsp;Installing on Macintosh</H3><!--SEC END -->

This section describes installation on pre-OS X machines. On OS X
Biopython can be installed using the UNIX instructions.<BR>
<BR>
Biopython code should work on Pre-OS X Macintoshes, using the MacPython distribution. I (Brad) am not a big Mac user, but have had good luck using several on the modules on the Macintosh.<BR>
<BR>
<!--TOC subsubsection Pre-Built Install-->

<H4><A NAME="htoc29">4.6.1</A>&nbsp;&nbsp;Pre-Built Install</H4><!--SEC END -->

Yair Benita has been kind enough prepare pre-compiled and ready to go
binaries for Macintosh machines. These distributions also contain the
required Biopython libraries, so you can get everything installed all at
once. <BR>
<BR>
These builds are available from the Biopython download page at
<CODE>.sit</CODE> files. You need to download this file and unpack it with
StuffIt Expander. This will create a <CODE>MacBiopython-version</CODE>
directory. You then just need to drag the contents of this directory to
a standard location python searches (something like 
<CODE>Macintosh HD::Python2.2::Lib::site-packages</CODE>) and it's all installed.<BR>
<BR>
<!--TOC subsubsection Non Pre-Built Installation-->

<H4><A NAME="htoc30">4.6.2</A>&nbsp;&nbsp;Non Pre-Built Installation</H4><!--SEC END -->

If you don't want to use the pre-built releases, you can get some basic
functionality from Biopython without compiling anything.
You need to download either the <CODE>biopython-version.tar.gz</CODE> or <CODE>biopython-version.zip</CODE> file from the download page, and unpack these. This can be done with tools such as Aladdin's Stuff-It expander. It will unpack into a directory called <CODE>biopython-version</CODE>. If you open up this directory, you will find the main directory of modules, called <CODE>Bio</CODE>. You should then open up your python installation (which should be in some place like <CODE>Macintosh HD::Python2.2</CODE>) to the directory <CODE>Lib::site-packages</CODE>, and copy the <CODE>Bio</CODE> directory there by dragging it. Bam! You're done! By default, <CODE>site-packages</CODE> is included in your <CODE>PYTHONPATH</CODE>, so you should be ready to use it.<BR>
<BR>
Some notes: Obviously this will not compile any of the C extensions in biopython. There are pure python implementations of all of these extensions, though, so you shouldn't need to worry about lack of functionality, only lack of speed. Jack Jansen (the MacPython god) has made patches to distutils which allow it to work on the Mac with the Metrowerks CodeWarrior compiler. I don't have this compiler (it costs money, oh no!), so I can't speak of how well it works. If anyone who codes more on the Mac has more information, I would be very happy to include it here.<BR>
<BR>
<!--TOC section Making sure everything worked-->

<H2><A NAME="htoc31">5</A>&nbsp;&nbsp;Making sure everything worked</H2><!--SEC END -->

<A NAME="sec:is_working"></A>
First, we'll just do a quick test to make sure Biopython is installed correctly. The most important thing is that python can find the biopython installation. Biopython installs into top level <CODE>Bio</CODE>, <CODE>Martel</CODE> and <CODE>BioSQL</CODE> directories, so you'll want to make sure these directories are located in a directory specified 
in your<CODE> $PYTHONPATH</CODE> environmental variable. If you used the default install, this shouldn't be a problem, but if not, you'll need to set the <CODE>PYTHONPATH</CODE> with something like <CODE>export PYTHONPATH = $PYTHONPATH':/directory/where/you/put/Biopython'</CODE> (on UNIX). Now that we think we are ready, fire up your python interpreter and follow along with the following code:
<PRE>
$ python
Python 2.4 (#1, Dec  5 2004, 20:47:03)
[GCC 3.3.3] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt; from Bio.Seq import Seq
&gt;&gt;&gt; from Bio.Alphabet.IUPAC import unambiguous_dna
&gt;&gt;&gt; new_seq = Seq('GATCAGAAG', unambiguous_dna)
&gt;&gt;&gt; new_seq[0:2]
Seq('GA', IUPACUnambiguousDNA())
&gt;&gt;&gt; from Bio import Translate
&gt;&gt;&gt; translator = Translate.unambiguous_dna_by_name["Standard"]
&gt;&gt;&gt; translator.translate(new_seq)
Seq('DQK', HasStopCodon(IUPACProtein(), '*'))
&gt;&gt;&gt;
</PRE>
If this worked properly, then it looks like Biopython is in a happy place where python can find it, so now you might want to do some more rigorous tests. The <CODE>Tests</CODE> directory inside the distribution contains a number of tests you can run to make sure all of the different parts of biopython are working. These should all work just by running <CODE>python test_WhateverTheTestIs.py</CODE>. <BR>
<BR>
You can also run all of the tests using a nice graphical interface supplied by using PyUnit. To do this, you just need to be in the installation directory and type:
<PRE>
python setup.py test
</PRE>
This should start up a Tk based graphical user interface (or default to the command line if you don't have Tkinter installed), which you can run the tests from. You can also run them by typing <CODE>python run_tests.py</CODE> in the Tests directory.<BR>
<BR>
If you've made it this far, you've gotten Biopython installed and running.
Congratulations!<BR>
<BR>
<!--TOC section Notes for installing with non-administrator permissions-->

<H2><A NAME="htoc32">6</A>&nbsp;&nbsp;Notes for installing with non-administrator permissions</H2><!--SEC END -->

Although I mentioned above that I wouldn't go much into installing in
non-root directories, if you are stuck installing
Biopython and it's dependencies into your home directory here are a
few notes and tricks to keep you going:
<UL><LI>
 The Numeric package will take a little trickery to get it to be
 able to import. It will install a <CODE>Numeric</CODE> directory and
 <CODE>Numeric.pth</CODE> file into <CODE>your_dir/lib/python</CODE>. The problem
 is that the <CODE>Numeric.pth</CODE> file only works when located in the main
 python <CODE>site-packages</CODE> directory. The solution, fortunately, is
 simple -- you need to make sure <CODE>your_dir/lib/python</CODE> is in the
 <CODE>PYTHONPATH</CODE> and then need to make a <CODE>__init__.py</CODE> file in
 the <CODE>Numeric</CODE> directory (an empty file or a file with just a
 comment is fine, as long as it exists). Then <CODE>import Numeric</CODE>
 should work.<BR>
<BR>
<LI>Building some C modules, such as <CODE>Bio.Cluster</CODE> require that 
 the Numeric include files (normally installed in
 <CODE>your_dir/include/python/Numeric</CODE>) be available. If the
 compiler can't find these directories you'll normally get an error
 like:
<PRE>
Bio/Cluster/clustermodule.c:2: Numeric/arrayobject.h: No such file or directory
    </PRE>
 Followed by a long messy list of syntax errors. To fix this, you'll
 have to edit the <CODE>setup.py</CODE> file to let it know where the
 include directories are located. Look for the line in
 <CODE>setup.py</CODE> that looks like:
<PRE>
    include_dirs=["Bio/Cluster"]
    </PRE>
 and adjust it so that it includes the include directory where the
 numeric libraries were installed:
<PRE>
    include_dirs=["Bio/Cluster", "your_dir/include/python"]
    </PRE>
 Then you should be able to install everything happily.</UL>
Yes, it's a bit of a mess installing lots of packages in non-standard
locations. The best solution is to talk with your friendly system
administrator and get them to assist with the installation of at least
the required packages (they are generally quite useful for any python
install) before going ahead with Biopython installation.<BR>
<BR>
<!--HTMLFOOT-->
<!--ENDHTML-->
<!--FOOTER-->
<HR SIZE=2>
<BLOCKQUOTE><EM>This document was translated from L<sup>A</sup>T<sub>E</sub>X by
</EM><A HREF="http://pauillac.inria.fr/~maranget/hevea/index.html"><EM>H<FONT SIZE=2><sup>E</sup></FONT>V<FONT SIZE=2><sup>E</sup></FONT>A</EM></A><EM>.
</EM></BLOCKQUOTE>
</BODY>
</HTML>
