# Copyright 2000 by Jeffrey Chang.  All rights reserved.
# This code is part of the Biopython distribution and governed by its
# license.  Please see the LICENSE file that should have been included
# as part of this package.

"""Code to work with GenBank -- http://www.ncbi.nlm.nih.gov/

Classes:
Iterator              Iterate through a file of GenBank entries
Dictionary            Access a GenBank file using a dictionary interface.
ErrorFeatreParser     Catch errors caused during parsing.
FeatureParser         Parse GenBank data in Seq and SeqFeature objects.
RecordParser          Parse GenBank data into a Record object.
NCBIDictionary        Access GenBank using a dictionary interface.

_BaseGenBankConsumer  A base class for GenBank consumer that implements
                      some helpful functions that are in common between
                      consumers.
_FeatureConsumer      Create SeqFeature objects from info generated by
                      the Scanner
_RecordConsumer       Create a GenBank record object from Scanner info.
_PrintingConsumer     A debugging consumer.

_Scanner              Set up a Martel based GenBank parser to parse a record.

ParserFailureError    Exception indicating a failure in the parser (ie.
                      scanner or consumer)
LocationParserError   Exception indiciating a problem with the spark based
                      location parser.

Functions:
index_file            Get a GenBank file ready to be used as a Dictionary.
search_for            Do a query against GenBank.
download_many         Download many GenBank records.

"""
__all__ = [
    'LocationParser',
    'Record',
    'genbank_format',
    ]

# other Biopython stuff
from Bio.ParserSupport import AbstractConsumer
import utils


class Dictionary:
    """Allow a GenBank file to be accessed using a dictionary interface.
    """
    __filename_key = '__filename'
    def __init__(self, index_file, parser = None):
        """Initialize and open up a GenBank dictionary.

        Arguments:
        o index_file - The name of the file that servers as the dictionary
        index. You need to use the index_file function in this module
        to creat this index.
        o parser - An optional argument specifying a parser object that
        the records should be run through before returning the output. If
        parser is None then the unprocessed contents of the file will be
        returned.
        """
        from Bio import Index
        self._index = Index.Index(index_file)
        self._handle = open(self._index[Dictionary.__filename_key])
        self._parser = parser

    def __len__(self):
        return len(self._index)

    def __getitem__(self, key):
        """Retrieve an item from the dictionary.
        """
        from Bio import File
        # print "keys:", self._index.keys()
        # get the location of the record of interest in the file
        start, len = self._index[key]
        # print "start:", start, "len:", len

        # read through and get the data from the file
        self._handle.seek(start)
        data = self._handle.read(len)
        # print "data:", data

        # run the data through the parser if one is specified
        if self._parser is not None:
            return self._parser.parse(File.StringHandle(data))

        return data

    def __getattr__(self, name):
        return getattr(self._index, name)

    def keys(self):
        """Provide valid keys from the index.

        If keys is just called on the index (using getattr) then
        the index will return internal values such as '__filename' which
        we just don't want to see. This just strips these values out
        before returning.
        """
        all_keys = [key for key in self._index.keys() if not key.startswith('__')]

        return all_keys
        
class Iterator:
    """Iterator interface to move over a file of GenBank entries one at a time.
    """
    def __init__(self, handle, parser = None, has_header = 0):
        """Initialize the iterator.

        Arguments:
        o handle - A handle with GenBank entries to iterate through.
        o parser - An optional parser to pass the entries through before
        returning them. If None, then the raw entry will be returned.
        o has_header - Deprecated, headers are now autodetected.
        """
        from Martel import RecordReader
        from Bio import File
        if isinstance(handle, File.UndoHandle):
            _handle = handle
        else:
            _handle = File.UndoHandle(handle)
        # skip ahead until we find first record
        while _handle.peekline().find("LOCUS") < 0:
            _handle.readline()

        self._reader = RecordReader.StartsWith(_handle, "LOCUS")          
        self._parser = parser

    def next(self):
        """Return the next GenBank record from the handle.

        Will return None if we ran out of records.
        """
        from Bio import File
        data = self._reader.next()

        if self._parser is not None:
            if data:
                return self._parser.parse(File.StringHandle(data))

        return data

class ParserFailureError(Exception):
    """Failure caused by some kind of problem in the parser.
    """
    pass

class LocationParserError(Exception):
    """Could not Properly parse out a location from a GenBank file.
    """
    pass

class ErrorParser:
    """Parse GenBank files and attempt to catch errors.

    This is just a small wrapper class around a passed parser which
    catches errors that occur in the parser.

    When errors occur, we'll raise a ParserFailureError, which returns
    (hopefully helpful) information about where the error occured
    """
    def __init__(self, parser, bad_file_handle = None):
        """Initialize an ErrorFeatureParser.

        Arguments:
        o parser - The actual parser to use in parsing the records.
        o bad_file_handle - A handle to write problem GenBank files to
        If None, the files will not be saved.
        """
        self._bad_file_handle = bad_file_handle

        self._parser = parser

    def parse(self, handle):
        """Parse the specified handle.
        """
        import Martel
        from Bio import File
        record = handle.read()

        try:
            return self._parser.parse(File.StringHandle(record))
        # deal with any problems as Parser Errors
        # XXX Watch out for changes in exceptions, as noted in Andrew's
        # comments in the Martel code.
        except (Martel.Parser.ParserException,
                Martel.Parser.ParserPositionException):
            if self._bad_file_handle:
                self._bad_file_handle.write(record)

            raise ParserFailureError("Could not parse record %s" %
                                     self._parser._consumer.data.id)
        # SystemExits often come from location parsing with spark
        except LocationParserError, msg:
            raise ParserFailureError("Could not parse location %s in record %s"
                                     % (msg, self._parser._consumer.data.id))
                                                          
class FeatureParser:
    """Parse GenBank files into Seq + Feature objects.
    """
    def __init__(self, debug_level = 0, use_fuzziness = 1, 
                 feature_cleaner = utils.FeatureValueCleaner()):
        """Initialize a GenBank parser and Feature consumer.

        Arguments:
        o debug_level - An optional argument that species the amount of
        debugging information Martel should spit out. By default we have
        no debugging info (the fastest way to do things), but if you want
        you can set this as high as two and see exactly where a parse fails.
        o use_fuzziness - Specify whether or not to use fuzzy representations.
        The default is 1 (use fuzziness).
        o feature_cleaner - A class which will be used to clean out the
        values of features. This class must implement the function 
        clean_value. GenBank.utils has a "standard" cleaner class, which
        is used by default.
        """
        self._scanner = _Scanner(debug_level)
        self.use_fuzziness = use_fuzziness
        self._cleaner = feature_cleaner

    def parse(self, handle):
        """Parse the specified handle.
        """
        self._consumer = _FeatureConsumer(self.use_fuzziness, 
                                          self._cleaner)
        self._scanner.feed(handle, self._consumer)
        return self._consumer.data

class RecordParser:
    """Parse GenBank files into Record objects
    """
    def __init__(self, debug_level = 0):
        """Initialize the parser.

        Arguments:
        o debug_level - An optional argument that species the amount of
        debugging information Martel should spit out. By default we have
        no debugging info (the fastest way to do things), but if you want
        you can set this as high as two and see exactly where a parse fails.
        """
        self._scanner = _Scanner(debug_level)

    def parse(self, handle):
        """Parse the specified handle into a GenBank record.
        """
        self._consumer = _RecordConsumer()
        self._scanner.feed(handle, self._consumer)
        return self._consumer.data

class _BaseGenBankConsumer(AbstractConsumer):
    """Abstract GenBank consumer providing useful general functions.

    This just helps to eliminate some duplication in things that most
    GenBank consumers want to do.
    """
    # Special keys in GenBank records that we should remove spaces from
    # For instance, \translation keys have values which are proteins and
    # should have spaces and newlines removed from them. This class
    # attribute gives us more control over specific formatting problems.
    remove_space_keys = ["translation"]

    def __init__(self):
        pass

    def _split_keywords(self, keyword_string):
        """Split a string of keywords into a nice clean list.
        """
        # process the keywords into a python list
        if keyword_string[-1] == '.':
            keywords = keyword_string[:-1]
        else:
            keywords = keyword_string
        keyword_list = keywords.split(';')
        clean_keyword_list = [x.strip() for x in keyword_list]
        return clean_keyword_list

    def _split_accessions(self, accession_string):
        """Split a string of accession numbers into a list.
        """
        # first replace all line feeds with spaces
        accession = accession_string.replace("\n", " ")

        accession_list = accession.split(' ')
        clean_accession_list = [x.strip() for x in accession_list]

        return clean_accession_list

    def _split_taxonomy(self, taxonomy_string):
        """Split a string with taxonomy info into a list.
        """
        if taxonomy_string[-1] == '.':
            tax_info = taxonomy_string[:-1]
        else:
            tax_info = taxonomy_string
        tax_list = tax_info.split(';')
        new_tax_list = []
        for tax_item in tax_list:
            new_items = tax_item.split("\n")
            new_tax_list.extend(new_items)
        while '' in new_tax_list:
            new_tax_list.remove('')
        clean_tax_list = [x.strip() for x in new_tax_list]

        return clean_tax_list

    def _clean_location(self, location_string):
        """Clean whitespace out of a location string.

        The location parser isn't a fan of whitespace, so we clean it out
        before feeding it into the parser.
        """
        import string
        location_line = location_string
        for ws in string.whitespace:
            location_line = location_line.replace(ws, '')

        return location_line

    def _remove_newlines(self, text):
        """Remove any newlines in the passed text, returning the new string.
        """
        # get rid of newlines in the qualifier value
        newlines = ["\n", "\r"]
        for ws in newlines:
            text = text.replace(ws, "")

        return text

    def _normalize_spaces(self, text):
        """Replace multiple spaces in the passed text with single spaces.
        """
        # get rid of excessive spaces
        text_parts = text.split(" ")
        text_parts = filter(None, text_parts)
        return ' '.join(text_parts)

    def _remove_spaces(self, text):
        """Remove all spaces from the passed text.
        """
        return text.replace(" ", "")

    def _convert_to_python_numbers(self, start, end):
        """Convert a start and end range to python notation.

        In GenBank, starts and ends are defined in "biological" coordinates,
        where 1 is the first base and [i, j] means to include both i and j.

        In python, 0 is the first base and [i, j] means to include i, but
        not j. 

        So, to convert "biological" to python coordinates, we need to 
        subtract 1 from the start, and leave the end and things should
        be converted happily.
        """
        new_start = start - 1
        new_end = end

        return new_start, new_end

class _FeatureConsumer(_BaseGenBankConsumer):
    """Create a SeqRecord object with Features to return.

    Attributes:
    o use_fuzziness - specify whether or not to parse with fuzziness in
    feature locations.
    o feature_cleaner - a class that will be used to provide specialized
    cleaning-up of feature values.
    """
    def __init__(self, use_fuzziness, feature_cleaner = None):
        from Bio.SeqRecord import SeqRecord
        _BaseGenBankConsumer.__init__(self)
        self.data = SeqRecord(None, id = None)

        self._use_fuzziness = use_fuzziness
        self._feature_cleaner = feature_cleaner

        self._seq_type = ''
        self._seq_data = []
        self._current_ref = None
        self._cur_feature = None
        self._cur_qualifier_key = None
        self._cur_qualifier_value = None

    def locus(self, locus_name):
        """Set the locus name is set as the name of the Sequence.
        """
        self.data.name = locus_name

    def size(self, content):
        pass

    def residue_type(self, type):
        """Record the sequence type so we can choose an appropriate alphabet.
        """
        self._seq_type = type

    def data_file_division(self, division):
        self.data.annotations['data_file_division'] = division

    def date(self, submit_date):
        self.data.annotations['date'] = submit_date 

    def definition(self, definition):
        """Set the definition as the description of the sequence.
        """
        self.data.description = definition

    def accession(self, acc_num):
        """Set the accession number as the id of the sequence.

        If we have multiple accession numbers, the first one passed is
        used.
        """
        new_acc_nums = self._split_accessions(acc_num)

        # if we haven't set the id information yet, add the first acc num
        if self.data.id is None:
            if len(new_acc_nums) > 0:
                self.data.id = new_acc_nums[0]

    def nid(self, content):
        self.data.annotations['nid'] = content

    def pid(self, content):
        self.data.annotations['pid'] = content

    def version(self, version_id):
        """Set the version to overwrite the id.

        Since the verison provides the same information as the accession
        number, plus some extra info, we set this as the id if we have
        a version.
        """
        self.data.id = version_id

    def db_source(self, content):
        self.data.annotations['db_source'] = content.rstrip()

    def gi(self, content):
        self.data.annotations['gi'] = content

    def keywords(self, content):
        self.data.annotations['keywords'] = self._split_keywords(content)

    def segment(self, content):
        self.data.annotations['segment'] = content

    def source(self, content):
        if content[-1] == '.':
            source_info = content[:-1]
        else:
            source_info = content
        self.data.annotations['source'] = source_info

    def organism(self, content):
        self.data.annotations['organism'] = content

    def taxonomy(self, content):
        self.data.annotations['taxonomy'] = self._split_taxonomy(content)
        
    def reference_num(self, content):
        """Signal the beginning of a new reference object.
        """
        from Bio.SeqFeature import Reference
        # if we have a current reference that hasn't been added to
        # the list of references, add it.
        if self._current_ref is not None:
            self.data.annotations['references'].append(self._current_ref)
        else:
            self.data.annotations['references'] = []

        self._current_ref = Reference()

    def reference_bases(self, content):
        """Attempt to determine the sequence region the reference entails.

        Possible types of information we may have to deal with:
        
        (bases 1 to 86436)
        (sites)
        (bases 1 to 105654; 110423 to 111122)
        1  (residues 1 to 182)
        """
        # first remove the parentheses or other junk
        ref_base_info = content[1:-1]

        all_locations = []
        # parse if we've got 'bases' and 'to'
        if ref_base_info.find('bases') != -1 and \
            ref_base_info.find('to') != -1:
            # get rid of the beginning 'bases'
            ref_base_info = ref_base_info[5:]
            locations = self._split_reference_locations(ref_base_info)
            all_locations.extend(locations)
        elif (ref_base_info.find("residues") >= 0 and
              ref_base_info.find("to") >= 0):
            residues_start = ref_base_info.find("residues")
            # get only the information after "residues"
            ref_base_info = ref_base_info[(residues_start + len("residues ")):]
            locations = self._split_reference_locations(ref_base_info)
            all_locations.extend(locations)

        # make sure if we are not finding information then we have
        # the string 'sites' or the string 'bases'
        elif (ref_base_info == 'sites' or
              ref_base_info.strip() == 'bases'):
            pass
        # otherwise raise an error
        else:
            raise ValueError("Could not parse base info %s in record %s" %
                             (ref_base_info, self.data.id))

        self._current_ref.location = all_locations

    def _split_reference_locations(self, location_string):
        """Get reference locations out of a string of reference information
        
        The passed string should be of the form:

            1 to 20; 20 to 100

        This splits the information out and returns a list of location objects
        based on the reference locations.
        """
        from Bio import SeqFeature
        # split possibly multiple locations using the ';'
        all_base_info = location_string.split(';')

        new_locations = []
        for base_info in all_base_info:
            start, end = base_info.split('to')
            new_start, new_end = \
              self._convert_to_python_numbers(int(start.strip()),
                                              int(end.strip()))
            this_location = SeqFeature.FeatureLocation(new_start, new_end)
            new_locations.append(this_location)
        return new_locations

    def authors(self, content):
        self._current_ref.authors = content

    def consrtm(self, content):
        self._current_ref.consrtm = content

    def title(self, content):
        self._current_ref.title = content

    def journal(self, content):
        self._current_ref.journal = content

    def medline_id(self, content):
        self._current_ref.medline_id = content

    def pubmed_id(self, content):
        self._current_ref.pubmed_id = content

    def remark(self, content):
        self._current_ref.comment = content

    def comment(self, content):
        self.data.annotations['comment'] = "\n".join(content)

    def features_line(self, content):
        """Get ready for the feature table when we reach the FEATURE line.
        """
        self.start_feature_table()

    def start_feature_table(self):
        """Indicate we've got to the start of the feature table.
        """
        # make sure we've added on our last reference object
        if self._current_ref is not None:
            self.data.annotations['references'].append(self._current_ref)
            self._current_ref = None

    def _add_feature(self):
        """Utility function to add a feature to the SeqRecord.

        This does all of the appropriate checking to make sure we haven't
        left any info behind, and that we are only adding info if it
        exists.
        """
        if self._cur_feature:
            # if we have a left over qualifier, add it to the qualifiers
            # on the current feature
            self._add_qualifier()

            self._cur_qualifier_key = ''
            self._cur_qualifier_value = ''
            self.data.features.append(self._cur_feature)
            
    def feature_key(self, content):
        from Bio import SeqFeature
        # if we already have a feature, add it on
        self._add_feature()

        # start a new feature
        self._cur_feature = SeqFeature.SeqFeature()
        self._cur_feature.type = content

        # assume positive strand to start with if we have DNA. The
        # complement in the location will change this later.
        if self._seq_type.find("DNA") >= 0:
            self._cur_feature.strand = 1

    def location(self, content):
        """Parse out location information from the location string.

        This uses Andrew's nice spark based parser to do the parsing
        for us, and translates the results of the parse into appropriate
        Location objects.
        """
        from Bio.GenBank import LocationParser
        # --- first preprocess the location for the spark parser
        
        # we need to clean up newlines and other whitespace inside
        # the location before feeding it to the parser.
        # locations should have no whitespace whatsoever based on the
        # grammer
        location_line = self._clean_location(content)

        # Older records have junk like replace(266,"c") in the
        # location line. Newer records just replace this with
        # the number 266 and have the information in a more reasonable
        # place. So we'll just grab out the number and feed this to the
        # parser. We shouldn't really be losing any info this way.
        if location_line.find('replace') != -1:
            comma_pos = location_line.find(',')
            location_line = location_line[8:comma_pos]
        
        # feed everything into the scanner and parser
        try:
            parse_info = \
                       LocationParser.parse(LocationParser.scan(location_line))
        # spark raises SystemExit errors when parsing fails
        except SystemExit:
            raise LocationParserError(location_line)

        # print "parse_info:", repr(parse_info)
        
        # add the parser information the current feature
        self._set_location_info(parse_info, self._cur_feature)

    def _set_function(self, function, cur_feature):
        """Set the location information based on a function.

        This handles all of the location functions like 'join', 'complement'
        and 'order'.

        Arguments:
        o function - A LocationParser.Function object specifying the
        function we are acting on.
        o cur_feature - The feature to add information to.
        """
        from Bio import SeqFeature
        from Bio.GenBank import LocationParser
        assert isinstance(function, LocationParser.Function), \
               "Expected a Function object, got %s" % function
        
        if function.name == "complement":
            # mark the current feature as being on the opposite strand
            cur_feature.strand = -1
            # recursively deal with whatever is left inside the complement
            for inner_info in function.args:
                self._set_location_info(inner_info, cur_feature)
        # deal with functions that have multipe internal segments that
        # are connected somehow.
        # join and order are current documented functions.
        # one-of is something I ran across in old files. Treating it
        # as a sub sequence feature seems appropriate to me.
        # bond is some piece of junk I found in RefSeq files. I have
        # no idea how to interpret it, so I jam it in here
        elif (function.name == "join" or function.name == "order" or
              function.name == "one-of" or function.name == "bond"):
            self._set_ordering_info(function, cur_feature)
        elif (function.name == "gap"):
            assert len(function.args) == 1, \
              "Unexpected number of arguments in gap %s" % function.args
            # make the cur information location a gap object
            position = self._get_position(function.args[0].local_location)
            cur_feature.location = SeqFeature.PositionGap(position)
        else:
            raise ValueError("Unexpected function name: %s" % function.name)

    def _set_ordering_info(self, function, cur_feature):
        """Parse a join or order and all of the information in it.

        This deals with functions that order a bunch of locations,
        specifically 'join' and 'order'. The inner locations are
        added as subfeatures of the top level feature
        """
        from Bio import SeqFeature
        # for each inner element, create a sub SeqFeature within the
        # current feature, then get the information for this feature
        for inner_element in function.args:
            new_sub_feature = SeqFeature.SeqFeature()
            # inherit the type from the parent
            new_sub_feature.type = cur_feature.type 
            # add the join or order info to the location_operator
            cur_feature.location_operator = function.name
            new_sub_feature.location_operator = function.name
            # inherit references and strand from the parent feature
            new_sub_feature.ref = cur_feature.ref
            new_sub_feature.ref_db = cur_feature.ref_db
            new_sub_feature.strand = cur_feature.strand

            # set the information for the inner element
            self._set_location_info(inner_element, new_sub_feature)

            # now add the feature to the sub_features
            cur_feature.sub_features.append(new_sub_feature)

        # set the location of the top -- this should be a combination of
        # the start position of the first sub_feature and the end position
        # of the last sub_feature

        # these positions are already converted to python coordinates 
        # (when the sub_features were added) so they don't need to
        # be converted again
        feature_start = cur_feature.sub_features[0].location.start
        feature_end = cur_feature.sub_features[-1].location.end
        cur_feature.location = SeqFeature.FeatureLocation(feature_start,
                                                          feature_end)

    def _set_location_info(self, parse_info, cur_feature):
        """Set the location information for a feature from the parse info.

        Arguments:
        o parse_info - The classes generated by the LocationParser.
        o cur_feature - The feature to add the information to.
        """
        from Bio.GenBank import LocationParser
        # base case -- we are out of information
        if parse_info is None:
            return
        # parse a location -- this is another base_case -- we assume
        # we have no information after a single location
        elif isinstance(parse_info, LocationParser.AbsoluteLocation):
            self._set_location(parse_info, cur_feature)
            return
        # parse any of the functions (join, complement, etc)
        elif isinstance(parse_info, LocationParser.Function):
            self._set_function(parse_info, cur_feature)
        # otherwise we are stuck and should raise an error
        else:
            raise ValueError("Could not parse location info: %s"
                             % parse_info)

    def _set_location(self, location, cur_feature):
        """Set the location information for a feature.

        Arguments:
        o location - An AbsoluteLocation object specifying the info
        about the location.
        o cur_feature - The feature to add the information to.
        """
        # check to see if we have a cross reference to another accession
        # ie. U05344.1:514..741
        if location.path is not None:
            cur_feature.ref = location.path.accession
            cur_feature.ref_db = location.path.database
        # now get the actual location information
        cur_feature.location = self._get_location(location.local_location)

    def _get_location(self, range_info):
        """Return a (possibly fuzzy) location from a Range object.

        Arguments:
        o range_info - A location range (ie. something like 67..100). This
        may also be a single position (ie 27).

        This returns a FeatureLocation object.
        If parser.use_fuzziness is set at one, the positions for the
        end points will possibly be fuzzy.
        """
        from Bio import SeqFeature
        from Bio.GenBank import LocationParser
        # check if we just have a single base
        if not(isinstance(range_info, LocationParser.Range)):
            pos = self._get_position(range_info)
            # move the single position back one to be consistent with how
            # python indexes numbers (starting at 0)
            pos.position = pos.position  - 1
            return SeqFeature.FeatureLocation(pos, pos)
        # otherwise we need to get both sides of the range
        else:
            # get *Position objects for the start and end
            start_pos = self._get_position(range_info.low)
            end_pos = self._get_position(range_info.high)

            start_pos.position, end_pos.position = \
              self._convert_to_python_numbers(start_pos.position,
                                              end_pos.position)

            return SeqFeature.FeatureLocation(start_pos, end_pos)

    def _get_position(self, position):
        """Return a (possibly fuzzy) position for a single coordinate.

        Arguments:
        o position - This is a LocationParser.* object that specifies
        a single coordinate. We will examine the object to determine
        the fuzziness of the position.

        This is used with _get_location to parse out a location of any
        end_point of arbitrary fuzziness.
        """
        from Bio import SeqFeature
        from Bio.GenBank import LocationParser
        # case 1 -- just a normal number
        if (isinstance(position, LocationParser.Integer)):
            final_pos = SeqFeature.ExactPosition(position.val) 
        # case 2 -- we've got a > sign
        elif isinstance(position, LocationParser.LowBound):
            final_pos = SeqFeature.AfterPosition(position.base.val)
        # case 3 -- we've got a < sign
        elif isinstance(position, LocationParser.HighBound):
            final_pos = SeqFeature.BeforePosition(position.base.val)
        # case 4 -- we've got 100^101
        elif isinstance(position, LocationParser.Between):
            final_pos = SeqFeature.BetweenPosition(position.low.val,
                                                 position.high.val)
        # case 5 -- we've got (100.101)
        elif isinstance(position, LocationParser.TwoBound):
            final_pos = SeqFeature.WithinPosition(position.low.val,
                                                position.high.val)
        # case 6 -- we've got a one-of(100, 110) location
        elif isinstance(position, LocationParser.Function) and \
                        position.name == "one-of":
            # first convert all of the arguments to positions
            position_choices = []
            for arg in position.args:
                # we only handle AbsoluteLocations with no path
                # right now. Not sure if other cases will pop up
                assert isinstance(arg, LocationParser.AbsoluteLocation), \
                  "Unhandled Location type %r" % arg
                assert arg.path is None, "Unhandled path in location"
                position = self._get_position(arg.local_location)
                position_choices.append(position)
            final_pos = SeqFeature.OneOfPosition(position_choices)
        # if it is none of these cases we've got a problem!
        else:
            raise ValueError("Unexpected LocationParser object %r" %
                             position)

        # if we are using fuzziness return what we've got
        if self._use_fuzziness:
            return final_pos
        # otherwise return an ExactPosition equivalent
        else:
            return SeqFeature.ExactPosition(final_pos.location)

    def _add_qualifier(self):
        """Add a qualifier to the current feature without loss of info.

        If there are multiple qualifier keys with the same name we
        would lose some info in the dictionary, so we append a unique
        number to the end of the name in case of conflicts.
        """
        # if we've got a key from before, add it to the dictionary of
        # qualifiers
        if self._cur_qualifier_key:
            key = self._cur_qualifier_key
            value = self._cur_qualifier_value
            if self._feature_cleaner is not None:
                value = self._feature_cleaner.clean_value(key, value)
            # if the qualifier name exists, append the value
            if key in self._cur_feature.qualifiers:
                self._cur_feature.qualifiers[key].append(value)
            # otherwise start a new list of the key with its values
            else:
                self._cur_feature.qualifiers[key] = [value]

    def qualifier_key(self, content):
        """When we get a qualifier key, use it as a dictionary key.
        """
        # add a qualifier if we've got one
        self._add_qualifier()

        # remove the / and = from the qualifier
        qual_key = content.replace('/', '')
        qual_key = qual_key.replace('=', '')
        qual_key = qual_key.strip()
        
        self._cur_qualifier_key = qual_key
        self._cur_qualifier_value = ''
        
    def qualifier_value(self, content):
        # get rid of the quotes surrounding the qualifier if we've got 'em
        qual_value = content.replace('"', '')
        
        self._cur_qualifier_value = qual_value

    def contig_location(self, content):
        """Deal with a location of CONTIG information.
        """
        from Bio import SeqFeature
        # add a last feature if is hasn't been added,
        # so that we don't overwrite it
        self._add_feature()

        # make a feature to add the information to
        self._cur_feature = SeqFeature.SeqFeature()
        self._cur_feature.type = "contig"
        
        # now set the location on the feature using the standard
        # location handler
        self.location(content)
        # add the contig information to the annotations and get rid
        # of the feature to prevent it from being added to the feature table
        self.data.annotations["contig"] = self._cur_feature
        self._cur_feature = None

    def origin_name(self, content):
        pass

    def base_count(self, content):
        pass

    def base_number(self, content):
        pass

    def sequence(self, content):
        """Add up sequence information as we get it.

        To try and make things speedier, this puts all of the strings
        into a list of strings, and then uses string.join later to put
        them together. Supposedly, this is a big time savings
        """
        new_seq = content.replace(' ', '')
        new_seq = new_seq.upper()

        self._seq_data.append(new_seq)

    def record_end(self, content):
        """Clean up when we've finished the record.
        """
        from Bio import Alphabet
        from Bio.Alphabet import IUPAC
        from Bio.Seq import Seq
        # add the last feature in the table which hasn't been added yet
        self._add_feature()

        # add the sequence information
        # first, determine the alphabet
        # we default to an generic alphabet if we don't have a
        # seq type or have strange sequence information.
        seq_alphabet = Alphabet.generic_alphabet

        if self._seq_type:
            # mRNA is really also DNA, since it is actually cDNA
            if self._seq_type.find('DNA') != -1 or \
               self._seq_type.find('mRNA') != -1:
                seq_alphabet = IUPAC.ambiguous_dna
            # are there every really RNA sequences in GenBank?
            elif self._seq_type.find('RNA') != -1:
                seq_alphabet = IUPAC.ambiguous_rna
            elif self._seq_type == "PROTEIN":
                seq_alphabet = IUPAC.protein  # or extended protein?
            # work around ugly GenBank records which have circular or
            # linear but no indication of sequence type
            elif self._seq_type in ["circular", "linear"]:
                pass
            # we have a bug if we get here
            else:
                raise ValueError("Could not determine alphabet for seq_type %s"
                                 % self._seq_type)

        # now set the sequence
        sequence = "".join(self._seq_data)
        self.data.seq = Seq(sequence, seq_alphabet)

class _RecordConsumer(_BaseGenBankConsumer):
    """Create a GenBank Record object from scanner generated information.
    """
    def __init__(self):
        _BaseGenBankConsumer.__init__(self)
        import Record
        self.data = Record.Record()

        self._seq_data = []
        self._cur_reference = None
        self._cur_feature = None
        self._cur_qualifier = None

    def locus(self, content):
        self.data.locus = content

    def size(self, content):
        self.data.size = content

    def residue_type(self, content):
        self.data.residue_type = content

    def data_file_division(self, content):
        self.data.data_file_division = content

    def date(self, content):
        self.data.date = content

    def definition(self, content):
        self.data.definition = content

    def accession(self, content):
        new_accessions = self._split_accessions(content)
        self.data.accession.extend(new_accessions)

    def nid(self, content):
        self.data.nid = content

    def pid(self, content):
        self.data.pid = content

    def version(self, content):
        self.data.version = content

    def db_source(self, content):
        self.data.db_source = content.rstrip()

    def gi(self, content):
        self.data.gi = content

    def keywords(self, content):
        self.data.keywords = self._split_keywords(content)

    def segment(self, content):
        self.data.segment = content

    def source(self, content):
        self.data.source = content

    def organism(self, content):
        self.data.organism = content

    def taxonomy(self, content):
        self.data.taxonomy = self._split_taxonomy(content)

    def reference_num(self, content):
        """Grab the reference number and signal the start of a new reference.
        """
        # check if we have a reference to add
        if self._cur_reference is not None:
            self.data.references.append(self._cur_reference)

        self._cur_reference = Record.Reference()
        self._cur_reference.number = content

    def reference_bases(self, content):
        self._cur_reference.bases = content

    def authors(self, content):
        self._cur_reference.authors = content

    def consrtm(self, content):
        self._cur_reference.consrtm = content

    def title(self, content):
        self._cur_reference.title = content

    def journal(self, content):
        self._cur_reference.journal = content

    def medline_id(self, content):
        self._cur_reference.medline_id = content

    def pubmed_id(self, content):
        self._cur_reference.pubmed_id = content

    def remark(self, content):
        self._cur_reference.remark = content

    def comment(self, content):
        self.data.comment = "\n".join(content)

    def features_line(self, content):
        """Get ready for the feature table when we reach the FEATURE line.
        """
        self.start_feature_table()

    def start_feature_table(self):
        """Signal the start of the feature table.
        """
        # we need to add on the last reference
        if self._cur_reference is not None:
            self.data.references.append(self._cur_reference)

    def feature_key(self, content):
        """Grab the key of the feature and signal the start of a new feature.
        """
        # first add on feature information if we've got any
        self._add_feature()

        self._cur_feature = Record.Feature()
        self._cur_feature.key = content

    def _add_feature(self):
        """Utility function to add a feature to the Record.

        This does all of the appropriate checking to make sure we haven't
        left any info behind, and that we are only adding info if it
        exists.
        """
        if self._cur_feature is not None:
            # if we have a left over qualifier, add it to the qualifiers
            # on the current feature
            if self._cur_qualifier is not None:
                self._cur_feature.qualifiers.append(self._cur_qualifier)

            self._cur_qualifier = None
            self.data.features.append(self._cur_feature)

    def location(self, content):
        self._cur_feature.location = self._clean_location(content)

    def qualifier_key(self, content):
        # add on a qualifier if we've got one
        if self._cur_qualifier is not None:
            self._cur_feature.qualifiers.append(self._cur_qualifier)

        self._cur_qualifier = Record.Qualifier()
        self._cur_qualifier.key = content

    def qualifier_value(self, content):
        cur_content = self._remove_newlines(content)
        # remove all spaces from the value if it is a type where spaces
        # are not important
        for remove_space_key in self.__class__.remove_space_keys:
            if self._cur_qualifier.key.find(remove_space_key) >= 0:
                cur_content = self._remove_spaces(cur_content)
        self._cur_qualifier.value = self._normalize_spaces(cur_content)

    def base_count(self, content):
        self.data.base_counts = content

    def origin_name(self, content):
        self.data.origin = content

    def contig_location(self, content):
        """Signal that we have contig information to add to the record.
        """
        self.data.contig = self._clean_location(content) 

    def sequence(self, content):
        """Add sequence information to a list of sequence strings.

        This removes spaces in the data and uppercases the sequence, and
        then adds it to a list of sequences. Later on we'll join this
        list together to make the final sequence. This is faster than
        adding on the new string every time.
        """
        new_seq = content.replace(' ', '')
        self._seq_data.append(new_seq.upper())

    def record_end(self, content):
        """Signal the end of the record and do any necessary clean-up.
        """
        # add together all of the sequence parts to create the
        # final sequence string
        self.data.sequence = "".join(self._seq_data)
        # add on the last feature
        self._add_feature()

def _strip_and_combine(line_list):
    """Combine multiple lines of content separated by spaces.

    This function is used by the EventGenerator callback function to
    combine multiple lines of information. The lines are first
    stripped to remove whitepsace, and then combined so they are separated
    by a space. This is a simple minded way to combine lines, but should
    work for most cases.
    """
    # first strip out extra whitespace
    stripped_line_list = [x.strip() for x in line_list]

    # now combine everything with spaces
    return ' '.join(stripped_line_list)

class _Scanner:
    """Start up Martel to do the scanning of the file.

    This initialzes the Martel based parser and connects it to a handler
    that will generate events for a Feature Consumer.
    """
    def __init__(self, debug = 0):
        """Initialize the scanner by setting up our caches.

        Creating the parser takes a long time, so we want to cache it
        to reduce parsing time.

        Arguments:
        o debug - The level of debugging that the parser should
        display. Level 0 is no debugging, Level 2 displays the most
        debugging info (but is much slower). See Martel documentation
        for more info on this.
        """
        import Martel
        import genbank_format
        # a listing of all tags we are interested in scanning for
        # in the MartelParser
        self.interest_tags = ["locus", "size", "residue_type",
                              "data_file_division", "date",
                              "definition", "accession", "nid", 
                              "pid", "version", "db_source",
                              "gi", "keywords", "segment",
                              "source", "organism",
                              "taxonomy", "reference_num",
                              "reference_bases", "authors", "consrtm", "title",
                              "journal", "medline_id", "pubmed_id",
                              "remark", "comment",
                              "features_line", "feature_key",
                              "location", "qualifier_key",
                              "qualifier_value", "origin_name",
                              "base_count", "base_number",
                              "sequence", "contig_location", "record_end"]

        # a listing of all tags which should be left alone with respect
        # to whitespace handles
        self.exempt_tags = ["comment"]

        # make a parser that returns only the tags we are interested in
        expression = Martel.select_names(genbank_format.record,
                                         self.interest_tags)
        self._parser = expression.make_parser(debug_level = debug)

    def feed(self, handle, consumer):
        """Feeed a set of data into the scanner.

        Arguments:
        o handle - A handle with the information to parse.
        o consumer - The consumer that should be informed of events.
        """
        # XML from python 2.0
        from xml.sax import handler
        from Bio.ParserSupport import EventGenerator
        self._parser.setContentHandler(EventGenerator(consumer,
                                                      self.interest_tags,
                                                      _strip_and_combine,
                                                      self.exempt_tags))
        self._parser.setErrorHandler(handler.ErrorHandler())

        self._parser.parseFile(handle)

def index_file_db(genbank_file, db_name, db_directory,
                  identifier = "locus", aliases = ["accession"],
                  keywords = [], always_index = 0):
    """Index a GenBank file into a database for quick loading.

    WARNING: This is very experimental and subject to change.
    It requires the use of Andrew Dalke's mindy.

    This is very similar to index_file, but uses a database instead
    of a flat file to store the information about the genbank_file.

    Arguments:

    o genbank_file - The GenBank formatted file that we want to index.

    o db_name - The name of the database to create. This name will allow you
    to retrieve the file later.

    o db_directory - The directory where the database information should be
    stored.

    o identifier - The primary identifier used to store records in the file
    under. This will be used for retrieving them later.

    o aliases - Secondary identifiers that point to the record. These can
    be used for searching if a primary identifier is not found. This is
    useful for GenBank since we'll index by a single identifier (the LOCUS
    identifier by default) but might want to search by some other
    identifier.

    o keywords - More advanced Mindy features that I'm not positive
    how to make full use of right now.

    o always_index - A flag indicating whether or not to index a file even
    if the file appears not to have changed. By default, the function will
    try to skip indexing if it thinks the file hasn't changed.
    """
    import os
    import genbank_format
    try:
        from mindy import mindy_index, mindy_search
    except ImportError:
        raise SystemExit("You must have mindy installed:\n" +
                         "http://www.biopython.org/~dalke/mindy-0.1.tar.gz")

    # try to skip the indexing if everything seems up to date
    if not(always_index):
        if os.path.exists(os.path.join(db_directory, db_name)):
            # load up the database and see if the file size is the same
            search_db = mindy_search.mindy_open(db_directory, db_name)

            if "file_sizes" in search_db.mindy_data:
                file_size = search_db.mindy_data["file_sizes"][genbank_file]

                if file_size == os.path.getsize(genbank_file):
                    print "File already indexed. Skipping...."
                    return

    if not(os.path.exists(db_directory)):
        os.makedirs(db_directory)

    mindy_db = mindy_index.create(db_directory, db_name)
    mindy_db.use_filename(genbank_file)

    indexer = mindy_index.SimpleIndexer(mindy_db, "genbank_record", identifier,
                                        aliases, keywords)
    gb_format = genbank_format.record_format

    # -- don't use this optimization right now, it causes genbank_format
    # to be reset so that future calls to it don't return all of the item
    # of interest
    #if hasattr(indexer, "_wanted_elements"):
    #    gb_format = Martel.select_names(genbank_format.record_format,
    #                                    indexer._wanted_elements)

    parser = gb_format.make_parser()
    parser.setContentHandler(indexer)

    mindy_db.use_filename(genbank_file)
    parser.parseFile(open(genbank_file, "rb"))
    
class MindyDictionary:
    """Access a GenBank file using a dictionary interface, though a Mindy DB.

    WARNING: This is very experimental and subject to change.
    It requires the use of Andrew Dalke's mindy.

    This is the Dictionary interface to use after you create an index
    database using the function index_file_db.
    """
    def __init__(self, db_name, db_directory, parser = None):
        """Initialize and open up a GenBank dictionary.

        Arguments:
        
        o db_name - The name of the database we should retrieve information
        from.

        o db_directory - The location of the database specified in db_name.
        
        o parser - An optional argument specifying a parser object that
        the records should be run through before returning the output. If
        parser is None then the unprocessed contents of the file will be
        returned.
        """
        try:
            from mindy import mindy_search
        except ImportError:
            raise SystemExit("You must have mindy installed:\n" +
                          "http://www.biopython.org/~dalke/mindy-0.1.tar.gz")

        self._search = mindy_search.mindy_open(db_directory, db_name)
        self._parser = parser

    def __len__(self):
        return len(self._search.identifiers)

    def __getitem__(self, key):
        """Retrieve an item from the indexed file.

        The key can be either a primary identifier or an alias. The lookup
        will first try to get the file via the primary identifier, and if
        it can't do this, will subsequently try to get it through the
        aliases to these keys. If the aliases are ambigous, an error will
        be raised.

        Most of the time I find it easiest to search by aliases (the GenBank
        accession numbers), but YMMV.
        """
        from Bio import File
        try:
            data = self._search[key]
        except KeyError:
            ids = self._search.aliases.get(key, [])
            
            if len(ids) == 0:
                raise KeyError("No records found for key %s" % key)
            elif len(ids) != 1:
                raise KeyError("Multiple records found for key %s" % key)
            else:
                data = self._search[ids[0]]
            
        # run the data through the parser if one is specified
        if self._parser is not None:
            return self._parser.parse(File.StringHandle(data))

        return File.StringHandle(data)

    def __getattr__(self, name):
        return getattr(self._index, name)

    def keys(self):
        """Provide all identifiers for the current database.
        """
        return self._search.identifiers.keys()

    def aliases(self):
        """Provide all aliases in the current database.
        """
        return self._search.aliases.keys()
           
def index_file(genbank_file, index_file, rec_to_key = None):
    """Index a GenBank file to prepare it for use as a dictionary.

    Arguments:
    o genbank_file - The name of the GenBank file to be index.
    o index_name - The name of the index file which will be created.
    o rec_to_key - A function object which, when called with a GenBank
    record object, will return a key to be used for the record. If no
    function is specified, then the accession numbers will be used as
    the keys.
    """
    import os
    from Bio import Index
    if not os.path.exists(genbank_file):
        raise ValueError("%s does not exist" % genbank_file)

    index = Index.Index(index_file, truncate = 1)
    index[Dictionary._Dictionary__filename_key] = genbank_file

    gb_iter = Iterator(open(genbank_file), parser = RecordParser())
    while 1:
        start = gb_iter._reader.positions[gb_iter._reader.index]
        rec = gb_iter.next()
        length = gb_iter._reader.positions[gb_iter._reader.index] - start

        # when we run out of records, stop looping
        if rec is None:
            break

        # if we have a function to get a key, use it, otherwise use the default
        if rec_to_key is not None:
            key = rec_to_key(rec)
        else:
            # if we have accession numbers in the record use them
            if len(rec.accession) >= 1:
                key = rec.accession[0]
            else:
                key = None

        # check for problems with the key
        if key is None:
            raise KeyError("Empty sequence key produced")
        elif index.has_key(key):
            raise KeyError("Duplicate key %s found" % key)

        # finally apply the information to the key
        index[key] = start, length

class NCBIDictionary:
    """Access GenBank using a read-only dictionary interface.

    Methods:
    
    """
    def __init__(self, database='sequences', format="gb", delay=5.0,
                 parser=None):
        """NCBIDictionary([database][, delay][, parser])

        Create a new Dictionary to access GenBank.  Valid values for
        database are 'genome', 'nucleotide', 'protein', 'popset', and
        'sequences'.  delay is the number of seconds to wait between
        each query (5 default).  parser is an optional parser object
        to change the results into another form.  If unspecified, then
        the raw contents of the file will be returned.

        """
        from Bio.WWW import RequestLimiter
        self.parser = parser
        self.limiter = RequestLimiter(delay)
        self.database = database
        # JTC: This is broken.  Why is it changing the value of format?
        if format:
            self.format = format
        elif self.database == 'nucleotide':
            self.format = 'gb'
        elif self.database == 'protein' or self.database == 'popset':
            self.format = 'gp'
        else:
            self.format = 'native'

    def __len__(self):
        raise NotImplementedError, "GenBank contains lots of entries"
    def clear(self):
        raise NotImplementedError, "This is a read-only dictionary"
    def __setitem__(self, key, item):
        raise NotImplementedError, "This is a read-only dictionary"
    def update(self):
        raise NotImplementedError, "This is a read-only dictionary"
    def copy(self):
        raise NotImplementedError, "You don't need to do this..."
    def keys(self):
        raise NotImplementedError, "You don't really want to do this..."
    def items(self):
        raise NotImplementedError, "You don't really want to do this..."
    def values(self):
        raise NotImplementedError, "You don't really want to do this..."
    
    def has_key(self, id):
        """S.has_key(id) -> bool"""
        try:
            self[id]
        except KeyError:
            return 0
        return 1

    def get(self, id, failobj=None):
        try:
            return self[id]
        except KeyError:
            return failobj
        raise "How did I get here?"

    def __getitem__(self, id):
        """S.__getitem__(id) -> object

        Return the GenBank entry.  id is the GenBank ID (gi) of the
        entry.  Raises a KeyError if there's an error.
        
        """
        from Bio.WWW import NCBI
        # First, check to see if enough time has passed since my
        # last query.
        self.limiter.wait()
        
        try:
            handle = NCBI.efetch(
                self.database, rettype=self.format, id=id)
        except IOError, x:
            # raise a KeyError instead of an IOError
            # XXX I really should distinguish between a real IOError and
            # if the id is not in the database.
            raise KeyError, x

        # If the id is not in the database, I get a message like:
        # 'GenPept does not exist for GI "433174"\012\012' 
        line = handle.peekline()
        if line.find('does not exist') >= 0:
            raise KeyError, line
        # if there is a problem with the server, we'll get back a line like:
        # Please try again later. Server error  for GI "7212005"
        elif line.find("Please try again later.") >= 0:
            raise KeyError, line
        # if a sequence has been withdrawn, we'll get a line like:
        # The sequence has been intentionally withdrawn : GI "9993999"
        elif line.find("intentionally withdrawn") >= 0:
            raise KeyError, line
        # sometimes the intentianally withdrawn sequences just
        # return and empty sting
        elif line == '':
            raise KeyError, "Empty information returned."
        elif line.lower().find('html') >= 0:
            raise KeyError, "I unexpectedly got back html-formatted data."
        # Parse the record if a parser was passed in.
        if self.parser is not None:
            return self.parser.parse(handle)
        return handle.read()

def search_for(search, database='nucleotide',
               reldate=None, mindate=None, maxdate=None,
               batchsize=100, delay=2, callback_fn=None,
               start_id=0, max_ids=None):
    """search_for(search[, reldate][, mindate][, maxdate]
    [, batchsize][, delay][, callback_fn][, start_id][, max_ids]) -> ids

    Search GenBank and return a list of the GenBank identifiers (gi's)
    that match the criteria.  search is the search string used to
    search the database.  Valid values for database are 'genome',
    'nucleotide', 'protein', 'popset', and 'sequences'.  reldate is
    the number of dates prior to the current date to restrict the
    search.  mindate and maxdate are the dates to restrict the search,
    e.g. 2002/01/01.  batchsize specifies the number of ids to return
    at one time.  By default, it is set to 10000, the maximum.  delay
    is the number of seconds to wait between queries (default 2).
    callback_fn is an optional callback function that will be called
    and passed a gi as results are retrieved.  start_id specifies the
    index of the first id to retrieve and max_ids specifies the
    maximum number of id's to retrieve.

    XXX NCBI's esearch script does not seem to respect the date
    parameters.  Please let me know if you figure out how to get it to
    work.

    """
    import re
    import sgmllib
    from Bio.WWW import RequestLimiter
    from Bio.WWW import NCBI
    
    class ResultParser(sgmllib.SGMLParser):
        # Parse the ID's out of the XML-formatted page that PubMed
        # returns.  The format of the page is:
        # [...]
        #    <Id>...</Id>
        # [...]
        def __init__(self):
            sgmllib.SGMLParser.__init__(self)
            self.ids = []
            self.in_id = 0
        def start_id(self, attributes):
            self.in_id = 1
        def end_id(self):
            self.in_id = 0
        _not_pmid_re = re.compile(r'\D')
        def handle_data(self, data):
            if not self.in_id:
                return
            # If data is just whitespace, then ignore it.
            data = data.strip()
            if not data:
                return
            # Everything here should be a PMID.  Check and make sure
            # data really is one.  A PMID should be a string consisting
            # of only integers.  Should I check to make sure it
            # meets a certain minimum length?
            if self._not_pmid_re.search(data):
                raise SyntaxError, \
                      "I expected an ID, but %s doesn't look like one." % \
                      repr(data)
            self.ids.append(data)

    params = {
        'db' : database,
        'term' : search,
        'reldate' : reldate,
        'mindate' : mindate,
        'maxdate' : maxdate
        }
    for k, v in params.items():
        if v is None:
            del params[k]

    limiter = RequestLimiter(delay)
    ids = []
    while max_ids is None or len(ids) < max_ids:
        parser = ResultParser()
        
        # Check to make sure enough time has passed before my
        # last search.  If not, then wait.
        limiter.wait()

        start = start_id + len(ids)
        max = batchsize
        if max_ids is not None and max > max_ids - len(ids):
            max = max_ids - len(ids)

        params['retstart'] = start
        params['retmax'] = max
        h = NCBI.esearch(**params)
        parser.feed(h.read())
        ids.extend(parser.ids)
        if callback_fn is not None:
            # Call the callback function with each of the new ID's.
            for id in parser.ids:
                callback_fn(id)
        if len(parser.ids) < max or not parser.ids:  # no more id's to read
            break
    return ids

def download_many(ids, callback_fn, database='nucleotide',
                  broken_fn=None, delay=120.0, faildelay=5.0,
                  batchsize=500, parser=None):
    """download_many(ids, callback_fn[, broken_fn][, delay][, faildelay][, batchsize])

    Download many records from GenBank.  ids is a list of gis or
    accessions.  Each time a record is downloaded, callback_fn is
    called with the text of the record.  broken_fn is an optional
    function that is called with the id of records that were not able
    to be downloaded.  delay is the number of seconds to wait between
    requests.  batchsize is the number of records to request each
    time.

    """
    from Bio.WWW import NCBI
    from Bio.WWW import RequestLimiter
    from Bio import File
    # parser is an undocumented parameter that allows people to
    # specify an optional parser to handle each record.  This is
    # dangerous because the results may be malformed, and exceptions
    # in the parser may disrupt the whole download process.
    if batchsize > 500 or batchsize < 1:
        raise ValueError, "batchsize must be between 1 and 500"

    if database == 'nucleotide':
        format = 'gb'
    elif database == 'protein' or database == 'popset':
        format = 'gp'
    else:
        format = 'native'

    limiter = RequestLimiter(delay)
    current_batchsize = batchsize
    
    # Loop until all the ids are processed.  We want to process as
    # many as possible with each request.  Unfortunately, errors can
    # occur.  Some id may be incorrect, or the server may be
    # unresponsive.  In addition, one broken id out of a list of id's
    # can cause a non-specific error.  Thus, the strategy I'm going to
    # take, is to start by downloading as many as I can.  If the
    # request fails, I'm going to half the number of records I try to
    # get.  If there's only one more record, then I'll report it as
    # broken and move on.  If the request succeeds, I'll double the
    # number of records until I get back up to the batchsize.
    nsuccesses = 0
    while ids:
        if current_batchsize > len(ids):
            current_batchsize = len(ids)
        
        id_str = ','.join(ids[:current_batchsize])

        # Make sure enough time has passed before I do another query.
        if not nsuccesses:
            limiter.wait(faildelay)
        else:
            limiter.wait()
        try:
            # If one or more of the id's are broken, this will raise
            # an IOError.
            handle = NCBI.efetch(
                db=database, id=id_str, retmode='text', rettype=format)

            # I'm going to check to make sure PubMed returned the same
            # number of id's as I requested.  If it didn't then I'm going
            # to raise an exception.  This could take a lot of memory if
            # the batchsize is large.
            results = handle.read()
            iter = Iterator(File.StringHandle(results))
            num_ids = 0
            while iter.next() is not None:
                num_ids = num_ids + 1
            if num_ids != current_batchsize:
                raise IOError
            handle = File.StringHandle(results)
        except IOError:   # Query did not work.
            if current_batchsize == 1:
                # There was only 1 id in the query.  Report it as
                # broken and move on.
                id = ids.pop(0)
                if broken_fn is not None:
                    broken_fn(id)
            else:
                # I don't know which one is broken.  Try again with
                # fewer id's.
                current_batchsize = current_batchsize / 2
            nsuccesses = 0
            continue
        nsuccesses = nsuccesses + 1

        # Iterate through the results and pass the records to the
        # callback.
        iter = Iterator(handle, parser)
        idnum = 0
        while 1:
            rec = iter.next()
            if rec is None:
                break
            callback_fn(ids[idnum], rec)
            idnum = idnum + 1

        ids = ids[current_batchsize:]

        # If I'm not downloading the maximum number of articles,
        # double the number for next time.
        if nsuccesses >= 2 and current_batchsize < batchsize:
            current_batchsize = current_batchsize * 2
            if current_batchsize > batchsize:
                current_batchsize = batchsize
